File: app/src/main/java/com/dpis/module/applist/AppListFilter.java
package com.dpis.module.applist;

import com.dpis.module.fonts.FontApplyMode;


import java.util.Locale;

public final class AppListFilter {
    public enum Tab {
        ALL_APPS,
        CONFIGURED_APPS
    }

    private AppListFilter() {
    }

    public static boolean matches(String query, Tab tab, String label, String packageName,
                           boolean systemApp, boolean inScope, Integer viewportWidthDp,
                           Integer fontScalePercent, String fontMode, String typefaceId,
                           boolean appSpecificConfigActive, boolean configured, boolean installed,
                           AppListFilterState state) {
        return matches(query, tab, label, packageName, systemApp, inScope, viewportWidthDp,
                fontScalePercent, fontMode, typefaceId, null, true,
                appSpecificConfigActive, configured, installed, state);
    }

    /** Compatibility overload for callers that do not carry the persisted hook-chain summary. */
    public static boolean matches(String query,
                           Tab tab,
                           String label,
                           String packageName,
                           boolean systemApp,
                           boolean inScope,
                           Integer viewportWidthDp,
                           Integer fontScalePercent,
                           String fontMode,
                           String typefaceId,
                           boolean dpisEnabled,
                           boolean appSpecificConfigActive,
                           boolean configured,
                           boolean installed,
                           AppListFilterState state) {
        return matches(query, tab, label, packageName, systemApp, inScope, viewportWidthDp,
                fontScalePercent, fontMode, typefaceId, null, dpisEnabled, appSpecificConfigActive,
                configured, installed, state);
    }

    public static boolean matches(String query,
                           Tab tab,
                           String label,
                           String packageName,
                           boolean systemApp,
                           boolean inScope,
                           Integer viewportWidthDp,
                           Integer fontScalePercent,
                           String fontMode,
                           String typefaceId,
                           String fontHookDomainsRaw,
                           boolean dpisEnabled,
                           boolean appSpecificConfigActive,
                           boolean configured,
                           boolean installed,
                           AppListFilterState state) {
        String normalizedQuery = query == null ? "" : query.trim().toLowerCase(Locale.ROOT);
        if (!normalizedQuery.isEmpty()) {
            String normalizedLabel = label.toLowerCase(Locale.ROOT);
            String normalizedPackage = packageName.toLowerCase(Locale.ROOT);
            if (!normalizedLabel.contains(normalizedQuery)
                    && !normalizedPackage.contains(normalizedQuery)) {
                return false;
            }
        }
        boolean fontConfigured = fontScalePercent != null
                && FontApplyMode.isEnabled(FontApplyMode.normalize(fontMode));
        boolean typefaceConfigured = typefaceId != null && !typefaceId.isBlank();
        boolean anyFontConfigured = fontConfigured || typefaceConfigured;
        boolean hookConfigured = fontHookDomainsRaw != null && !fontHookDomainsRaw.trim().isEmpty();
        boolean matchesTab = switch (tab) {
            case ALL_APPS -> installed;
            case CONFIGURED_APPS -> configured;
        };
        if (!matchesTab) {
            return false;
        }

        AppListFilterState effectiveState = state != null
                ? state
                : AppListFilterState.noAdditionalConstraints();
        if (!effectiveState.allAppsSelected()
                && ((systemApp && !effectiveState.systemAppsSelected())
                || (!systemApp && !effectiveState.userAppsSelected()))) {
            return false;
        }
        if (effectiveState.injectedOnly() && !inScope) {
            return false;
        }
        if (effectiveState.disabledOnly() && dpisEnabled) return false;
        if (effectiveState.widthConfiguredOnly() && viewportWidthDp == null) {
            return false;
        }
        if (effectiveState.fontConfiguredOnly() && !anyFontConfigured) return false;
        if (effectiveState.typefaceConfiguredOnly() && !typefaceConfigured) return false;
        return !effectiveState.hookConfiguredOnly() || hookConfigured;
    }

}

File: app/src/main/java/com/dpis/module/applist/AppListFilter.kt
package com.dpis.module.applist

import com.dpis.module.fonts.FontApplyMode
import java.util.Locale

/** Pure matching policy for the app catalogue. */
object AppListFilter {
    enum class Tab {
        ALL_APPS,
        CONFIGURED_APPS,
    }

    @JvmStatic
    fun matches(
        query: String?,
        tab: Tab,
        label: String,
        packageName: String,
        systemApp: Boolean,
        inScope: Boolean,
        viewportWidthDp: Int?,
        fontScalePercent: Int?,
        fontMode: String?,
        typefaceId: String?,
        fontHookDomainsRaw: String?,
        dpisEnabled: Boolean,
        configured: Boolean,
        installed: Boolean,
        state: AppListFilterState,
    ): Boolean {
        val normalizedQuery = query.orEmpty().trim().lowercase(Locale.ROOT)
        if (normalizedQuery.isNotEmpty()) {
            val normalizedLabel = label.lowercase(Locale.ROOT)
            val normalizedPackage = packageName.lowercase(Locale.ROOT)
            if (!normalizedLabel.contains(normalizedQuery)
                && !normalizedPackage.contains(normalizedQuery)
            ) {
                return false
            }
        }

        if (tab == Tab.ALL_APPS && !installed) return false
        if (tab == Tab.CONFIGURED_APPS && !configured) return false
        if (state.selectedAppTypes.isNotEmpty()) {
            val appType = if (systemApp) {
                AppListFilterState.AppType.SYSTEM
            } else {
                AppListFilterState.AppType.USER
            }
            if (appType !in state.selectedAppTypes) return false
        }

        val fontConfigured = fontScalePercent != null
                && FontApplyMode.isEnabled(FontApplyMode.normalize(fontMode))
        val typefaceConfigured = !typefaceId.isNullOrBlank()
        val hookConfigured = !fontHookDomainsRaw.isNullOrBlank()
        val anyFontConfigured = fontConfigured || typefaceConfigured

        // Configuration chips describe included capabilities: selecting several
        // capabilities broadens the result to apps matching any selected filter.
        return state.selectedConfigurationFilters.isEmpty()
                || state.selectedConfigurationFilters.any { filter ->
            when (filter) {
                AppListFilterState.ConfigurationFilter.INJECTED -> inScope
                AppListFilterState.ConfigurationFilter.DISABLED -> !dpisEnabled
                AppListFilterState.ConfigurationFilter.VIEWPORT -> viewportWidthDp != null
                AppListFilterState.ConfigurationFilter.FONT -> anyFontConfigured
                AppListFilterState.ConfigurationFilter.TYPEFACE -> typefaceConfigured
                AppListFilterState.ConfigurationFilter.HOOK -> hookConfigured
                AppListFilterState.ConfigurationFilter.ALL -> true
            }
        }
    }
}
